### PR TITLE
Support PgAudit with Babelfish and addition of PgAudit github action

### PR DESCRIPTION
**Pull request by pranavJ23**
### Description

[Describe what this change achieves]
Extension changes PR link: https://github.com/babelfish-for-postgresql/babelfish_extensions/pull/2903

Postgres event triggers can call pg_event_trigger_ddl_commands which in turn does a syscache lookup for the object that fired the event trigger. If the event is create babelfish temp table or table variable then the syscahe lookup will fail since ENR sys table scan is only allowed when dialect is TSQL but when executing pg_event_trigger_ddl_commands() dialect will be PSQL. As a fix we will temporarily switch the dialect to TSQL when doing a syscahe lookup inside pg_event_trigger_ddl_commands()
 
 
### Issues Resolved

[List any issues this PR will resolve]

### Task
 [BABEL-5269], [BABEL-5233]

Signed-off-by: Pranav Jain [pranavjc@amazon.com](mailto:pranavjc@amazon.com)
 
### Check List

- [x] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is under the terms of the PostgreSQL license, and grant any person obtaining a copy of the contribution permission to relicense all or a portion of my contribution to the PostgreSQL License solely to contribute all or a portion of my contribution to the PostgreSQL open source project.

For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/babelfish-for-postgresql/postgresql_modified_for_babelfish/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: src/backend/commands/event_trigger.c [hunk 1]
```diff
@@ -1895,7 +1896,16 @@ pg_event_trigger_ddl_commands(PG_FUNCTION_ARGS)
 					 * rather than failing.  This can happen for example with
 					 * some subcommand combinations of ALTER TABLE.
 					 */
-					identity = getObjectIdentity(&addr, true);
+					if(cache_look_ddl_event_trigger_hook)
+					{
+						char temp_identity;
```
Comment (tanscorpio7) [inline src/backend/commands/event_trigger.c:1901]: this should be `char *`.

Comment (tanscorpio7) [inline src/backend/commands/event_trigger.c]: `temp_identity` should be of type char *.

File: src/backend/commands/event_trigger.c [hunk 2]
```diff
@@ -1895,7 +1896,16 @@ pg_event_trigger_ddl_commands(PG_FUNCTION_ARGS)
...
+						(*cache_look_ddl_event_trigger_hook)(&addr, &temp_identity);
+						identity = &temp_identity;
```
Comment (tanscorpio7) [inline src/backend/commands/event_trigger.c]: No need to pass by reference, we can simply return the `identity` value using hook.

**Review by tanscorpio7 — CHANGES_REQUESTED** [review]

File: src/backend/commands/event_trigger.c [hunk 3]
```diff
@@ -1895,7 +1896,14 @@ pg_event_trigger_ddl_commands(PG_FUNCTION_ARGS)
 					 * rather than failing.  This can happen for example with
 					 * some subcommand combinations of ALTER TABLE.
 					 */
-					identity = getObjectIdentity(&addr, true);
+					if(cache_look_ddl_event_trigger_hook && getObjectClass(&addr) == OCLASS_CLASS)
```
Comment (tanscorpio7) [inline src/backend/commands/event_trigger.c]: Please move the check for object class to inside hook.

**Review by tanscorpio7 — CHANGES_REQUESTED** [review]

File: src/backend/commands/event_trigger.c (hunk 3 above)
Comment (tanscorpio7) [reply src/backend/commands/event_trigger.c]: @pranavJ23 ^

File: src/include/commands/event_trigger.h [hunk 4]
```diff
@@ -85,4 +85,7 @@ extern void EventTriggerCollectAlterTSConfig(AlterTSConfigurationStmt *stmt,
 											 Oid cfgId, Oid *dictIds, int ndicts);
 extern void EventTriggerCollectAlterDefPrivs(AlterDefaultPrivilegesStmt *stmt);
 
+typedef char* (*cache_look_ddl_event_trigger_hook_type) (ObjectAddress* address);
+extern PGDLLIMPORT cache_look_ddl_event_trigger_hook_type cache_look_ddl_event_trigger_hook;
```
Comment (tanscorpio7) [inline src/include/commands/event_trigger.h]: Would `pltsql_get_object_identity_event_trigger` be a more descriptive name ?

File: src/include/tcop/utility.h [hunk 5]
```diff
@@ -16,6 +16,7 @@
 
 #include "tcop/cmdtag.h"
 #include "tcop/tcopprot.h"
+#include "catalog/dependency.h"
```
Comment (tanscorpio7) [inline src/include/tcop/utility.h]: Should this be `catalog/objectaddress.h`

**Review by tanscorpio7 — CHANGES_REQUESTED** [review]

File: src/backend/commands/event_trigger.c [hunk 6]
```diff
@@ -1895,7 +1896,22 @@ pg_event_trigger_ddl_commands(PG_FUNCTION_ARGS)
...
 					 */
-					identity = getObjectIdentity(&addr, true);
+
+					/*
+					 * To support pgAudit extension we are adding a hook
+					 * For the case when we have the object type as 
+					 * table(temp table, table variable, hash table)
+					 * Along with hook not null, we will initialise 
+					 * the syscache lookup.
+					*/
```
Comment (shah-nirmit) [inline src/backend/commands/event_trigger.c]: nit: indentation 
Multiline comment style is
```
/*
 *
 */
```

File: src/backend/commands/event_trigger.c [hunk 7]
```diff
@@ -1895,7 +1896,20 @@ pg_event_trigger_ddl_commands(PG_FUNCTION_ARGS)
...
-					identity = getObjectIdentity(&addr, true);
+
+					/*
+					 * When PgAudit extension is enabled we want to initialise 
+					 * the syscache lookup for babelfish temp tables and table 
+					 * variable from postgres dialect.
+					 */
+					if (pltsql_get_object_identity_event_trigger_hook)
+					{
+						identity=(*pltsql_get_object_identity_event_trigger_hook)(&addr);
```
Comment (tanscorpio7) [inline src/backend/commands/event_trigger.c]: indentation

File: src/backend/commands/event_trigger.c [hunk 8]
```diff
@@ -1895,7 +1896,20 @@ pg_event_trigger_ddl_commands(PG_FUNCTION_ARGS)
 					 * rather than failing.  This can happen for example with
 					 * some subcommand combinations of ALTER TABLE.
 					 */
-					identity = getObjectIdentity(&addr, true);
+
+					/*
+					 * When PgAudit extension is enabled we want to initialise 
+					 * the syscache lookup for babelfish temp tables and table 
+					 * variable from postgres dialect.
+					 */
```
Comment (tanscorpio7) [inline src/backend/commands/event_trigger.c]: comment should be in hooks.c in extension

**Review by tanscorpio7 — APPROVED** [review]
Please get the code changes reviewed from @surendravishnoi also

File: src/backend/commands/event_trigger.c (hunk 8 above)
Comment (shah-nirmit) [reply src/backend/commands/event_trigger.c]: +1

**Review by shardgupta — APPROVED** [review]

**Pull request merged**